examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+FEED_FORWARD_DIM = 256
+NUM_HEADS = 3
+NUM_LAYERS = 2
+VOCAB_SIZE = 5000  # Limits parameters in model.
+
+"""
+## Load Tokenizer
+
+Here we download the vocabulary data and initialize
+`keras_nlp.tokenizers.WordPieceTokenizer`. WordPieceTokenizer is an efficient
+implementation of the WordPiece algorithm used by BERT and other models. It will strip,
+lower-case and do other un-reversable preprocessing operations.
+
+The vocabulary data we download is a list of sub-words sorted in descending order by
+frequency. We want to limit the vocabulary as much as possible, as we will see later on
+that it has a large affect on the number of model parameters. We also don't want to take
+too few vocabulary, or there would be too many OOV sub-words. 
+
+We strip the first 1000 vocabulary that are all of the form `"[unusedXX]"`, since they
+are all not useful for this text corpus and unneccesarily increases vocabulary size. Then
+we take the `VOCAB_SIZE` most frequent vocabularies after that. There are also some
+reserved tokens, so we want to make sure that `"[PAD]"` is at index 0 and `"[UNK]"` is in
+the vocabulary, since it is used for padding and unknown words respectively.
+"""
+
+# Download vocabulary data.
+vocab_file = keras.utils.get_file(
+    origin="https://storage.googleapis.com/tensorflow/keras-nlp/examples/bert/bert_vocab_uncased.txt",
+)
+# Get most frequent VOCAB_SIZE vocab tokens for faster/efficient training.
+vocab_list = []
+with open(vocab_file, "r") as f:
+    for line in f:
+        vocab_list.append(line.strip())
+
+# The first 1000 tokens are unused tokens that unneccesarily increases vocab size.
+vocab_list = ["[PAD]", "[UNK]", "[CLS]", "[SEP]", "[MASK]"] + vocab_list[1000:]
+vocab_list = vocab_list[:VOCAB_SIZE]
+
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=vocab_list,
+    sequence_length=SEQ_LEN + 1,  # +1 for slicing feature and label.
+)
+
+"""
+## Load Data
+
+Now, let's download the dataset! The SimpleBooks dataset has a small vocabulary, which
+makes it easier to fit a simple model on.
+"""
+
+keras.utils.get_file(
+    origin="https://dldata-public.s3.us-east-2.amazonaws.com/simplebooks.zip",
+    extract=True,
+)
+dir = os.path.expanduser("~/.keras/datasets/simplebooks/")
+
+# Load simplebooks-92 train set and filter out short lines.
+train_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/train.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+    .shuffle(buffer_size=256)
+)
+
+# Load simplebooks-92 validation set and filter out short lines.
+valid_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/valid.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+)
+
+"""
+## Tokenize Data
+
+We pre-process the dataset by tokenizing and splitting it into `features` and `labels`.
+"""
+
+
+def preprocess(inputs):
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
```

Comment:
should we add bos tokens here? then you could sample without seed text, and I think we would be a little closer to actual gpt

Reply:
edited